[DON'T MERGE] Prototype for SlowOperationDetector (Hogging Partition Thread PRD)

Donnerbart opened the pull request:
This is a prototype for the `SlowOperationDetector` which detects and logs slow operations in `OperationThreads`.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/monitor/LocalOperationStats.java:
```diff
@@ -0,0 +1,4 @@
+package com.hazelcast.monitor;
+
+public interface LocalOperationStats extends LocalInstanceStats {
```
Can you add javadoc?

pveentjer commented on hazelcast/src/main/java/com/hazelcast/monitor/MemberState.java:
```diff
@@ -47,6 +47,8 @@
 
     LocalMemoryStats getLocalMemoryStats();
 
+    LocalOperationStats getOperationStats();
```
Same here.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/BasicOperationScheduler.java:
```diff
@@ -505,22 +508,26 @@ private void processPriorityMessages() {
 
-        @edu.umd.cs.findbugs.annotations.SuppressWarnings({"VO_VOLATILE_INCREMENT" })
+        @edu.umd.cs.findbugs.annotations.SuppressWarnings({"VO_VOLATILE_INCREMENT"})
         private void process(Object task) {
```
I would set 'currentOperation' directly in the beginning of the method and slap a try/finally around it so it is unset in the finally clause.

Donnerbart replied:
Hmm, we wouldn't get rid of any existing try-catch-block and the whole method will be indented once more. Not sure if that will really improve the readability.

bilalyasar commented on hazelcast/src/main/java/com/hazelcast/monitor/impl/LocalOperationStatsImpl.java:
```diff
@@ -0,0 +1,55 @@
+package com.hazelcast.monitor.impl;
```
missing class hearders

Donnerbart replied:
Done, again comment won't disappear because noted line is unchanged.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/BasicOperationService.java:
```diff
@@ -145,11 +144,16 @@
+        this.asyncExecutor = executionService
+                .register(ExecutionService.ASYNC_EXECUTOR, coreSize, ASYNC_QUEUE_CAPACITY, ExecutorType.CONCRETE);
+
+        this.slowOperationDetector = new SlowOperationDetector(node, this.scheduler);
```
Try to inject concrete dependencies you use instead of Node. Now the whole things becomes only testable with a whole cluster up and running.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,171 @@
+/**
+ * Monitors the operation threads to find and collect data of long running operations.
+ */
+final class SlowOperationDetector extends Thread {
```
Design tip; don't extend from a thread for these things, but have an internal thread. It is an implementation detail and you don't want to leak it to the outside.

Especially when you have a single scanning thread, you can do something like this:

```
class SlowOperationDetector{

	void scan(){
		... actual scanning logic	
	}

	private class ScanThread extends Thread{
		public void run(){
			while(!shutdown){
				scan();			
			}		
		}	
	}
}
```
Now when you want to do a unit test, you don't call the start method but manually call the scan. You remember the story I told you about the axle? The scan method is the axle, and the thread is the engine that rotates that axle.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,171 @@
+        genericOperationThreads = scheduler.genericOperationThreads;
+        partitionOperationThreads = scheduler.partitionOperationThreads;
+
+        genericCurrentOperationData = new CurrentOperationData[genericOperationThreads.length];
```
I would do it like this:

genericCurrentOperationData = initGenericCurrentOperationData()

And like the init method take care of creating and filling the array. Now it is a bit smeared and makes the constructor longer than needs to be. The same things goes for the partition...array.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,171 @@
+        int purgeIntervalSeconds = Math.abs(node.groupProperties.SLOW_OPERATION_DETECTOR_LOG_PURGE_INTERVAL_SECONDS.getInteger());
+        int logLifetimeSeconds = Math.abs(node.groupProperties.SLOW_OPERATION_DETECTOR_LOG_RETENTION_SECONDS.getInteger());
+
+        slowOperationThresholdNanos = TimeUnit.MILLISECONDS.toNanos(thresholdMillis);
```
I would pull this out in a methods:

slowOperationTHesholdNanos = initSlowOperationThreadholdNannos()... 

And in that method your read out the group properties + do your conversion + math.abs stuff.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,171 @@
+    private long nowNanos;
+    private long nowMillis;
+
+    private volatile boolean running = true;
```
Minor, but I would call this shutdown. We do the same thing with other services like the OperationService and scheduler.

Donnerbart replied:
I think this is bad code style, since you create a negation in the main usage of the variable): "if not state do something"

In this particular case it may be quite easy to parse, but I've seen this approach with very nasty variable names, which gave you a very hard way to understand the meaning of code. Sometimes the variables started with "not", so you had "if not notState do something".

So I got really used to avoid negation of boolean variables in the main code path.

pveentjer replied:
Certainly a good argument. 

Normally I prefer such loops:

```
 private void doRun() {
            for (;;) {
                Object task;
                try {
                    task = workQueue.take();
                } catch (InterruptedException e) {
                    if (shutdown) {
                        return;
                    }
                    continue;
                }

                if (shutdown) {
                    return;
                }

                processPriorityMessages();
                process(task);
            }
        }

```

And this is where the shutdown is prefered..

pveentjer replied:
Anyhow.. it doesn't need to be changed

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,171 @@
+            try {
+                TimeUnit.NANOSECONDS.sleep(ONE_SECOND_IN_NANOS - (System.nanoTime() - nowNanos));
+            } catch (InterruptedException ignored) {
+                break;
```
You don't need to break. The wile loop will decide if the thread needs to stop. So I would just remove the break;

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,171 @@
+    private volatile boolean running = true;
+
+    public SlowOperationDetector(Node node, BasicOperationScheduler scheduler) {
+        super(SlowOperationDetector.class.getSimpleName());
```
You need to make use of the HazelcastThreadGroup; see other threads being created.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,171 @@
+
+        while (running) {
+            nowNanos = System.nanoTime();
+            nowMillis = System.currentTimeMillis();
```
Why do you need to have both??

Donnerbart replied:
The `System.nanoTime()` is to measure the duration of an operation, because it's the most accurate method we have. Since it has nothing to do with our real time `System.currentTimeMillis()` is used to calculate the `startedAt` of an operation (so we have a timestamp which a user can read in the management center).

It doesn't matter if the values are a bit off. But they are meant as now (especially for the `startedAt` calculation I think it is much easier to read than a start/begin without having an end).

pveentjer replied:
Hmmm.. I didn't think about adding time meta-data....

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,171 @@
+        }
+    }
+
+    private void checkForSlowOperations(BasicOperationScheduler.OperationThread[] operationThreads,
```
Personally I would call this method 'scan' since the purpose of this class is to detect slow running operations anyway. I would also pull out the content of the loop into a seperate method. It is getting a bit on the long side.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    private SlowOperationLog getOrCreateSlowOperationLog(Thread operationThread, final Object operation) {
+        StringBuilder stackTrace = new StringBuilder();
```
I need to think about this class because are going to create quite a lot of litter. We can reuse the StringBuilder btw; since only a single thread will be using it.

pveentjer replied:
I would call it stackTraceSb or something. I got a bit confused while reading the code.

Donnerbart replied:
The stack trace is just created if an operation is being slow for the first time. After that the `Invocation` reference is cached in the `CurrentOperationData`. So there shouldn't be more overhead with the stack traces as necessary.

pveentjer replied:
You are creating an StringBuffer per slow operation; that isn't needed. This thing can be recycled.

pveentjer replied:
Just for my understanding.

if an operation is running e.g. for 30 minutes and we scan every minute, how many stack traces are being created? One or 30?

pveentjer commented on hazelcast/src/test/java/com/hazelcast/spi/impl/SlowOperationDetectorTest.java:
```diff
@@ -0,0 +1,483 @@
+    }
+
+    @Test
+    public void testMultipleSlowEntryProcessorClasses() throws InterruptedException {
```
You are testing 2 things; one is the json stuff the other is the scanning behavior. I would not do that because it makes your test difficult to understand.

Donnerbart replied:
Note should be outdated but not hidden because not changed line was noted.

Donnerbart commented:
Fixed `equals()` issue with `LocalOperationStats` and added more missing file headers.

Donnerbart commented:
verify

Donnerbart commented:
Failed test should be unrelated.
```
Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 16.416 sec <<< FAILURE! - in com.hazelcast.client.ClientReconnectTest
testClientReconnectOnClusterDownWithEntryListeners(com.hazelcast.client.ClientReconnectTest)  Time elapsed: 8.592 sec  <<< ERROR!
java.lang.RuntimeException: com.hazelcast.client.HazelcastClientNotActiveException: Partition does not have owner. partitionId : 58
	at com.hazelcast.client.spi.impl.ClientInvocation.notify(ClientInvocation.java:177)
	at com.hazelcast.client.spi.impl.ClientInvocation.invoke(ClientInvocation.java:134)
	at com.hazelcast.client.spi.ClientProxy.invoke(ClientProxy.java:129)
	at com.hazelcast.client.proxy.ClientMapProxy.put(ClientMapProxy.java:361)
	at com.hazelcast.client.proxy.ClientMapProxy.put(ClientMapProxy.java:203)
	at com.hazelcast.client.ClientReconnectTest$4.run(ClientReconnectTest.java:107)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:483)
	at com.hazelcast.client.ClientReconnectTest.testClientReconnectOnClusterDownWithEntryListeners(ClientReconnectTest.java:104)
```

Donnerbart commented:
The latest changes in `OperationXXX` classes were quite hard to adept. We need a way to get the `OperationThread` from the `OperationHandler` to capture a stack trace of a slow operation. The actual implementation is not very elegant.

pveentjer commented:
Hi David, good point.

Let me think about how to do that. For the classic scheduler it is quite easy because each handler is handled by a single thread, but for the progressive scheduler it is much more complicated because partition operations can be executed by user threads and therefore the relationship is more complicated.

So probably I'll expose a method 'currentThread' or something.. you can use to get a stacktrace from.

pveentjer commented:
Hi David, see this branch:

https://github.com/hazelcast/hazelcast/pull/4654

I think this should resolve the issues you are having. Give me a ping of the documentation is not clear.

```
/*
 * Copyright (c) 2008-2013, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hazelcast.spi.impl.operationexecutor;

import com.hazelcast.nio.Packet;
import com.hazelcast.spi.Operation;

/**
 * The OperationRunner is responsible for the actual running of operations.
 * <p/>
 * So the {@link OperationExecutor} is responsible for 'executing' them (so finding a thread to run on), the actual work is done
 * by the {@link OperationRunner}. This separation of concerns makes the code a lot simpler and makes it possible to swap parts
 * of the system.
 * <p/>
 * Since HZ 3.5 there are multiple OperationRunner instances; each partition will have its own OperationRunner, but also
 * generic threads will have their own OperationRunners. Each OperationRunner exposes the Operation it is currently working
 * on and this makes it possible to hook on all kinds of additional functionality like detecting slow operations, sampling which
 * operations are executed most frequently, check if an operation is still running, etc etc.
 */
public abstract class OperationRunner {

    protected final int partitionId;
    protected volatile Object currentTask;
    private volatile Thread currentThread;

    public OperationRunner(int partitionId) {
        this.partitionId = partitionId;
    }

    public abstract void run(Packet packet) throws Exception;

    public abstract void run(Runnable task);

    public abstract void run(Operation task);

    /**
     * Returns the current task that is executing. This value could be null if no operation is executing.
     * <p/>
     * Value could be stale as soon as it is returned.
     * <p/>
     * This method is thread-safe; so the thread that executes a task will set/unset the current task,
     * any other thread in the system is allowed to read it.
     *
     * @return the current running task.
     */
    public final Object currentTask() {
        return currentTask;
    }

    /**
     * Sets the thread that is running this OperationRunner instance.
     * <p/>
     * This method should only be called from the {@link OperationExecutor} since this component is responsible for
     * executing operations on an OperationRunner.
     *
     * @param currentThread the current Thread. Can be called with 'null', clearing the currentThread field.
     */
    public final void setCurrentThread(Thread currentThread) {
        this.currentThread = currentThread;
    }

    /**
     * Get the thread that is currently running this OperationRunner instance.
     * <p/>
     * This value only has meaning when an Operation is running. It depends on the implementation if the field is unset
     * after an operation is executed or not. So it could be that a value is returned while no operation is running.
     * <p/>
     * For example, the {@link com.hazelcast.spi.impl.operationexecutor.classic.ClassicOperationExecutor} will never unset
     * this field since each OperationRunner is bound to a single OperationThread; so this field is initialized when the
     * OperationRunner is created.
     * <p/>
     * The returned value could be null. When it is null, currently no thread is running this OperationRunner.
     * <p/>
     * Recommended idiom for slow operation detection:
     * 1: First read the operation and store the reference.
     * 2: Then read the current thread and store the reference
     * 3: Later read the operation again. If the operation-instance is the same, it means that you have captured the right thread.
     * <p/>
     * Then you use this Thread to create a stack trace. It can happen that the stracktrace doesn't reflect the call-state the
     * thread had when the slow operation was detected. This could be solved by rechecking the currentTask after you have detected
     * the slow operation. BUt don't create a stacktrace before you do the first recheck of the operation because otherwise it
     * will cause a lot of overhead.
     *
     * @return the Thread that currently is running this OperationRunner instance.
     */
    public final Thread currentThread() {
        return currentThread;
    }

    /**
     * Returns the partitionId this OperationRunner is responsible for. If the partition id is smaller than 0,
     * it is either a generic or ad hoc OperationRunner.
     * <p/>
     * The value will never change for this OperationRunner instance.
     *
     * @return the partition id.
     */
    public final int getPartitionId() {
        return partitionId;
    }
}
```

Donnerbart commented:
Rebased with master. I did a little extra cleanup, e.g. sorting of init methods, fixed some comments etc.

I also pulled out constants for the `-2` and `-1` partition ids to `OperationRunner`, to get rid of the magic ids in the code (at least on server side and in tests). I didn't have a look at the client side so far. But I think we should have these values as constants, because `-2` and `-1` aren't very verbose (and at least `-1` is used in different contexts, e.g. default/disabled timeout values).

Maybe this is just useful for the `BasicXXX` or `ClassicXXX` implementations, but then we should pull out the `BasicOperationRunner` or make it public, so we can put and use the constants from there.

`OperationRunner.AD_HOC_PARTITION_ID` is the `-2` and `OperationRunner.GENERIC_PARTITION_ID` is the `-1` now.

pveentjer commented:
You should revert the -2/-1 stuff in the partiton runner. That is not needed to be known in the operation runner for the time being.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationexecutor/classic/ClassicOperationExecutor.java:
```diff
@@ -368,15 +365,15 @@ public void execute(Packet packet) {
     public void runOnCallingThread(Operation op) {
         if (op == null) {
-            throw new NullPointerException("op can't be null");
+            throw new NullPointerException("Operation can't be null");
```
Why did you make this change? The name of the argument should be listed in the exception, not the class. If you want to rename op --> operation, than that is fine to me.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationexecutor/classic/ClassicOperationExecutor.java:
```diff
@@ -439,18 +436,13 @@ private static void awaitTermination(OperationThread[] operationThreads) {
-                    .append(" pendingCount=").append(operationThread.scheduleQueue.size())
-                    .append('\n');
+        for (PartitionOperationThread operationThread : partitionOperationThreads) {
+            sb.append(operationThread.getName()).append(" processedCount=").append(operationThread.processedCount)
```
This is not an improvement. Please revert it.

Donnerbart replied:
This is done via the code formatter which is provided in the wiki and recommended to use. If you have complains about that, we should fix the code formatter.

pveentjer replied:
Perhaps there is an issue with your IDE. BUt when I select the original code and I do a reformat, Idea doesn't change it. So please revert it.

pveentjer replied:
I just loaded the HZ code formatter. And it does indeed f*ck up the code. Perhaps time to drop the current formatter and just use the default provided one by IDEA.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationexecutor/classic/ClassicOperationExecutor.java:
```diff
@@ -439,18 +436,13 @@ private static void awaitTermination(OperationThread[] operationThreads) {
-            sb.append(operationThread.getName())
-                    .append(" processedCount=").append(operationThread.processedCount).append('\n');
+        for (GenericOperationThread operationThread : genericOperationThreads) {
+            sb.append(operationThread.getName()).append(" processedCount=").append(operationThread.processedCount).append('\n');
```
Same here.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationexecutor/classic/ClassicOperationExecutor.java:
```diff
@@ -176,17 +166,24 @@ private ResponseThread initResponseThread() {
 
-            GenericOperationThread operationThread = new GenericOperationThread(
-                    threadName, threadId, genericScheduleQueue,
+            GenericOperationThread operationThread = new GenericOperationThread(threadName, threadId, genericScheduleQueue,
```
This doesn't improve readability.

pveentjer commented on hazelcast/src/test/java/com/hazelcast/spi/impl/BasicBackPressureServiceTest.java:
```diff
@@ -142,7 +143,7 @@ public void whenNormalOperation(boolean partitionSpecific, boolean remoteCall) {
         } else {
             op = new GenericOperation();
-            op.setPartitionId(-1);
+            op.setPartitionId(OperationRunner.GENERIC_PARTITION_ID);
```
Revert all the constants you have added the OperationRunner.

pveentjer commented on hazelcast/src/test/java/com/hazelcast/spi/impl/SlowOperationDetectorTest.java:
```diff
@@ -0,0 +1,589 @@
+        }
+    }
+
+    private static class NestedSlowOperationOnDifferentThreadPools extends AbstractOperation {
```
I don't understand the meaning of this class, because you are not doing anything with threadpools. So what is this operation doing?

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,214 @@
+        SlowOperationLog.Invocation invocation;
+    }
+
+    private class Scanner extends Thread {
```
I would call it ScanThread. So that it immediately clear that it is a thread. If something is a thread or not is quite important to make very obvious.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,214 @@
+                    if (purge(nowNanos, lastLogPurge)) {
+                        lastLogPurge = nowNanos;
+                    }
+
```
I would pull this into a method in the Scanner to keep the method nice and snappy.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,214 @@
+        }
+
+        public void run() {
+            while (running) {
```
Why is there a nested loop?

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,214 @@
+                }
+            }
+        }
+    }
```
To simplify testing, I would follow the following approach:

expose a method purge. This method is the one you are going to call from your tests.

expose a method start. This method starts a thread that will call the purge method. For testing, you don't call start; so there is no thread doing stuff you don't want it to do.

This way testing is a lot easier because you don't need to deal with that nasty thread.

```
 public void purge(){
    long nowNanos = System.nanoTime();
    long nowMillis = System.currentTimeMillis();

    scan(nowNanos, nowMillis, genericOperationRunners, genericCurrentOperationData);
    scan(nowNanos, nowMillis, partitionOperationRunners, partitionCurrentOperationData);

    if (purge(nowNanos, lastLogPurge)) {
        lastLogPurge = nowNanos;
    }

 }

 public void start(){
    new ScanThread(...).start();
 }


class ScanThread extends Thread{
...
 public void run() {
                lastLogPurge = System.nanoTime();

                while (running) {
                    
                    try {
                        TimeUnit.NANOSECONDS.sleep(ONE_SECOND_IN_NANOS - (System.nanoTime() - nowNanos));
                    } catch (Exception ignored) {
                        EmptyStatement.ignore(ignored);
                    }
                }
        }
}
```

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,214 @@
+            return true;
+        }
+        return false;
+    }
```
I personally prefer jumping out of a method as soon as you see there is nothing to do. This reduces the amount of nesting.

It will also reduce the npath complexity of your method.

```
  private boolean purge(long nowNanos, long lastLogPurge) {
        if (nowNanos - lastLogPurge <= logPurgeIntervalNanos) {
            return false;
        }
        
       for (SlowOperationLog log : slowOperationLogs.values()) {
                log.purgeInvocations(nowNanos, logRetentionNanos);
                if (log.isEmpty()) {
                    slowOperationLogs.remove(log.getStackTrace().hashCode());
                }
         }           
 }
```

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,214 @@
+        return TimeUnit.SECONDS.toNanos(groupProperty.getInteger());
+    }
+
+    private void scan(long nowNanos, long nowMillis, OperationRunner[] operationRunners,
```
This method is too big. One small reduction; pull out the content of the for loop into a separate method.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,214 @@
+                continue;
+            }
+
+            // create the stack trace once for this operation and cache the invocation for upcoming updates
```
Probably this can be pulled out into a method that deals with logging.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,214 @@
+        genericCurrentOperationData = initCurrentOperationData(genericOperationRunners);
+        partitionCurrentOperationData = initCurrentOperationData(partitionOperationRunners);
+
+        slowOperationThresholdNanos = getGroupPropertyMillis(groupProperties.SLOW_OPERATION_DETECTOR_THRESHOLD_MILLIS);
```
Bit confused here because you are gettingXXXMillies but actually they are nanos. Perhaps getMilliesToNanos(...)

This way it is clear that you are expecting to find a millies and converting it to nanos. Anyhow.. not extremely relevant.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,214 @@
+            SlowOperationLog.Invocation invocation = currentOperationData.invocation;
+            if (invocation != null) {
+                // update existing invocation to avoid creation of stack trace
+                invocation.duration = TimeUnit.NANOSECONDS.toMillis(duration);
```
Can you add the timeunit to duration, e.g. durationMs.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,214 @@
+            if (invocation != null) {
+                // update existing invocation to avoid creation of stack trace
+                invocation.duration = TimeUnit.NANOSECONDS.toMillis(duration);
+                invocation.lastAccess = nowNanos;
```
Also add timeout to lastAccess, e.g. lastAccessNs.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationLog.java:
```diff
@@ -0,0 +1,196 @@
+import static com.hazelcast.util.JsonUtil.getLong;
+import static com.hazelcast.util.JsonUtil.getString;
+
+public final class SlowOperationLog implements JsonSerializable {
```
Can you add a few lines of javadoc?

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationLog.java:
```diff
@@ -0,0 +1,196 @@
+
+    private final ConcurrentHashMap<Integer, Invocation> invocations = new ConcurrentHashMap<Integer, Invocation>();
+
+    private String operation;
```
My concurrency alarm went off here because SlowOperationLog is shared between threads afaik and these fields are not threadafe to use. But afaik the fields can't change, only they are mutable to deal with serialization, right?

If so, can you add a few lines of javadoc that these fields can be considered immutable.

Donnerbart replied:
Added a comment.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationLog.java:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    String getShortStackTrace() {
+        if (shortStackTrace == null) {
```
Hmmm.. shortTraceTrace isn't volatile.... 

So could it be racy?? I'm not completely sure if getShortStackTrace can be called by multiple threads or not, but you might have a publication issue on your hands here

Donnerbart replied:
It's just used by a single thread and is `transient` (so not part of the public JSON).

In the worst case another thread would create the same shortened stack trace. It's like the `String.hashCode()`. I added comment to make this clear.

pveentjer replied:
Regarding worst case:

You can apply the hash trick for an int because you get the guarantee that no 'out of thin air' value is presented. But for complex object graphs this trick doesn't work.

Donnerbart replied:
I still don't see an issue here. It's just a cached value to log a shorter version of the stack trace. It's not important at all. The only method accessing the value is package private and just used by the `SlowOperationDetector` thread.

The worst case is a superfluous String on the heap (not even a wrong value), which can only be created by a class within `com.hazelcast.spi.impl`.

pveentjer replied:
Creating the unwanted stacktrace isn't relevant. What is relevant is that one threads can sees a object graph in a totally 'impossible' state due to lack of a happens before relation.

It isn't an issue because you just explained that the method can only called by a single thread; but that wasn't obvious to me since it isn't documented.

Donnerbart replied:
I don't see that problem either. The original stack trace is always set in the constructor. So even if two threads would call that method, they would have the same data and would create the same value (just in different String instances). It should be the same behavior like `String.hashCode()` if called from multiple threads. What am I missing?

pveentjer replied:
I'm would not dare to do it and i know 1 or 2 things about the JMM. You won't get publication issues because the String has its own safe publication. But I don't know the official guarantees on pointer addresses from the JMM. I'm sure they are provided, but I don't see a clear reason to live on the edge unless you are Doug Lea.

Anyhow.. if this method is only called by a single thread, then I don't see any reason to waste more of my time on this discussion apart from you documenting this behavior.

Donnerbart replied:
The documentation is there since some commits. You don't see it in the marked snippet above, because you made the original note on a line which has not changed :)

pveentjer replied:
Just looked it up:

 Writes to and reads of references are always atomic, regardless of whether they are implemented as 32 or 64 bit values.

But I would not quickly make use of such an optimisation; especially not with documentation.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationLog.java:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    Invocation getOrCreateInvocation(int operationHashCode, long lastDurationNanos, long nowNanos, long nowMillis) {
+        totalInvocations++;
```
Also racy + publication issue?

Donnerbart replied:
Good catch, that one should be `volatile`.

pveentjer replied:
Also document that there is only a single thread calling this method and therefor you don't need do have an atomic increment.

Donnerbart replied:
How would it make a difference if multiple threads would read that value? As long as just a single threads changes it `volatile` should be sufficient. Seen this in multiple classes with a similar comment then mine.

pveentjer replied:
Reading the volatile value is fine. But the ++ is not atomic. I'm perfectly doing fine doing ++ on a volatile field, but then it needs documentation. And in your case this documentation is missing.

Donnerbart replied:
It's not missing. It is on the variable itself, like I saw it in all the other places where this pattern is used. Again, snippet above is not updated, because you commented on the usage, which was not changed.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationLog.java:
```diff
@@ -0,0 +1,196 @@
+
+        sb.append("SlowOperationLog{operation='").append(operation).append("', stackTrace='").append(stackTrace)
+          .append("', totalInvocations=").append(totalInvocations).append(", invocations='");
+
```
What do you think of the following layout:

```
sb.append("SlowOperationLog{operation='").append(operation)
	.append("', stackTrace='").append(stackTrace)   
	.append("', totalInvocations=").append(totalInvocations)
	.append(", invocations='");
```

This is a lot easier to read because you can now see what belongs to each other. Long chaining of .append(..) are imho a PITA.

Donnerbart replied:
The code formatter from the wiki always changes those lines. We should see if we can fix that settings, otherwise we have to live with those things or don't use it anymore (which would be a PITA, too).

pveentjer replied:
Can you:

1: revert the code formatting issue
2: fix the formatter in the wiki so it doesn't mess up this code.

The default intellij code formatter doesn't have any problem with this code.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/InternalOperationService.java:
```diff
@@ -65,6 +67,13 @@
+     *
+     * @return collection of long running operation logs.
+     */
+    Collection<SlowOperationLog> getSlowOperationLogs();
```
Add comment that one should not mess with the values e.g. make modifications.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java:
```diff
@@ -0,0 +1,214 @@
+        final String stackTraceString = stackTraceStringBuilder.toString();
+        stackTraceStringBuilder.setLength(0);
+
+        return ConcurrencyUtil.getOrPutIfAbsent(slowOperationLogs, stackTraceString.hashCode(),
```
You should cache this function. Because now you are going to create a new function instance for every getOrPutIfAbsent call.

Donnerbart replied:
It doesn't make sense to cache this, because it is just called once per operation.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/spi/impl/operationexecutor/classic/ClassicOperationExecutor.java:
```diff
@@ -148,27 +141,24 @@ private ResponseThread initResponseThread() {
+        // we need to assign the PartitionOperationThreads to all OperationRunners they own
+        for (int partitionId = 0; partitionId < partitionOperationRunners.length; partitionId++) {
+            int threadId = partitionId % threadCount;
+            partitionOperationRunners[partitionId].setCurrentThread(threads[threadId]);
```
Can you make this line less complicated?

e.g.

```
Thread thread = threads[threadId]
PartitionOperationRunner runner = runners[partitionId];
runner.setThread(thread)
```
This is a lot easier to read.

Donnerbart commented:
Just to be sure about this: You want to keep (especially) the `-1` in the code instead of using a constant? I wouldn't care if it would just be about the ad-hoc operations (edge-case), but the generic threads are commonly used and the `-1` looks like a magic number by the book.

Donnerbart commented:
I already wrote that I don't care where we would put those constants (the `OperationRunner` may be the wrong place for it, since the values are not used there). But I think it makes the code easier to read to have those constants instead of plain numbers.

Donnerbart commented:
I just split up the test files to make it easier to find the different test sub-categories.

pveentjer commented:
I don't care that much about introducing a constant for -1. I would not like to have a dependency on OperationRunner to store the -1, since only classes in the .impl package should mess around with this class.

But the -2 for ad hoc is just an implementation detail of the BasicOperationService so it should not leak to the outside (otherwise it would not be an implementation detail).

Donnerbart commented:
Okay, that makes sense. the ad-hoc is really internal, so I will move back that constant to the orginate class.

Do you know a better place for the GENERIC_PARTITION_ID contant? Maybe `Operation` itself? It is set there as a default and it's a normal class (so no abuse of the abstract `OperationRunner` anymore).

pveentjer commented:
Indeed Operation could be a good location for it.

Donnerbart commented on hazelcast/src/test/java/com/hazelcast/spi/impl/operationexecutor/classic/PartitionSpecificCallable.java:
```diff
@@ -30,8 +30,8 @@ public int getPartitionId() {
-        } catch (Throwable e) {
-
+        } catch (Throwable t) {
+            throwable = t;
```
I hope this change is okay. Just saw that `throwable is not set anywhere and it is private.

pveentjer replied:
THink so. otherwise the exception would need to be called 'ignored'.

pveentjer commented:
I have just merged this PR because it looks good. Lets try it in action. Do we have a stabilizer test to simulate slow behavior?

Donnerbart commented:
`com.hazelcast.stabilizer.tests.slow.SlowOperationMapTest`
